authorization/msi-auth-token-provider-jar/src/main/java/com/microsoft/azure/msiAuthTokenProvider/MSICredentials.java

```diff
@@ -0,0 +1,356 @@
+            default:
+                throw new IllegalArgumentException("unknown host type:" + hostType);
+        }
+    }
+
+    private MSIToken getTokenForAppService(String tokenAudience) throws IOException, ParseException {
+        String urlString = null;
+
+        if (this.configForAppService.msiEndpoint() == null || this.configForAppService.msiEndpoint().isEmpty()) {
+            //the web app does not have MSI set, return file not found
+            throw new FileNotFoundException("Managed identity not found/configured");
+        }
+
+        if (this.configForAppService.msiClientId() != null && !this.configForAppService.msiClientId().isEmpty()) {
+            urlString = String.format("%s?resource=%s&clientid=%s&api-version=2017-09-01", this.configForAppService.msiEndpoint(),
+                    tokenAudience == null ? this.configForAppService.resource() : tokenAudience,
+                    this.configForAppService.msiClientId());
+        } else {
+            urlString = String.format("%s?resource=%s&api-version=2017-09-01", this.configForAppService.msiEndpoint(),
+                    tokenAudience == null ? this.configForAppService.resource() : tokenAudience);
+        }
+        URL url = new URL(urlString);
+        HttpURLConnection connection = null;
+
+        try {
+            connection = (HttpURLConnection) url.openConnection();
+
+            connection.setRequestMethod("GET");
+            connection.setRequestProperty("Secret", this.configForAppService.msiSecret());
+            connection.setRequestProperty("Metadata", "true");
+
+            connection.connect();
+
+            InputStream stream = connection.getInputStream();
+            BufferedReader reader = new BufferedReader(new InputStreamReader(stream, "UTF-8"), 100);
+            String result = reader.readLine();
+
+            return getMsiTokenFromResult(result, HostType.APP_SERVICE);
+        } catch (IOException ioEx){
+            if  (ioEx.getMessage().contains("Server returned HTTP response code: 400 for URL")) {
+                throw new FileNotFoundException("Managed identity not found/configured");
+            } else throw ioEx;
+        } catch (Exception e){
+            if (e.getCause()!= null && e.getCause() instanceof SocketException && e.getCause().getMessage().contains("Permission denied: connect")) {
+                throw new FileNotFoundException("Managed identity not found/configured");
+            } else throw e;
+        } finally {
+            if (connection != null) {
+                connection.disconnect();
+            }
+        }
+    }
+
+    private MSIToken getTokenForVirtualMachineFromIMDSEndpoint(String tokenAudience) {
+        MSIToken token = null;
+
+            try {
+                token = retrieveTokenFromIDMSWithRetry(tokenAudience);
+            } catch (IOException exception) {
+                throw new RuntimeException(exception);
```

Comment:
Do we need this try-catch given `retrieveTokenFromIDMSWithRetry` has a try-catch for generic exception?

Reply:
yeah this is a relic from my older version of the code.. ll pull this out.